common/src/service/ExceptionHandlingHelper.cs

```diff
@@ -116,12 +117,56 @@ public static async Task<ErrorCode> GetExceptionCodeAsync(HttpResponseMessage re
             try
             {
                 IoTHubExceptionResult responseContent = JsonConvert.DeserializeObject<IoTHubExceptionResult>(responseContentStr);
-                Dictionary<string, string> messageFields = JsonConvert.DeserializeObject<Dictionary<string, string>>(responseContent.Message);
 
-                if (messageFields != null
-                    && messageFields.TryGetValue(CommonConstants.ErrorCode, out string errorCodeObj))
+                try
                 {
-                    errorCodeValue = Convert.ToInt32(errorCodeObj, CultureInfo.InvariantCulture);
+                    Dictionary<string, string> messageFields = JsonConvert.DeserializeObject<Dictionary<string, string>>(responseContent.Message);
+
+                    if (messageFields != null
+                        && messageFields.TryGetValue(CommonConstants.ErrorCode, out string errorCodeObj))
+                    {
+                        errorCodeValue = Convert.ToInt32(errorCodeObj, CultureInfo.InvariantCulture);
+                    }
+                }
+                catch (JsonReaderException ex)
+                {
+                    if (Logging.IsEnabled)
+                        Logging.Error(null, $"Failed to deserialize error message into a dictionary: {ex}. Message body: '{responseContentStr}.'");
+
+                    // In some scenarios, the error response string is a ';' delimited string with the service-returned error code.
+                    const char errorFieldsDelimiter = ';';
+                    string[] messageFields = responseContent.Message.Split(errorFieldsDelimiter);
+
+                    if (messageFields != null)
+                    {
+                        _ = messageFields
+                            .Any(field =>
```

Comment:
Any reason for using Linq instead of foreach here? The result of .Any is not being so true/false does not matter but we always return it as true, which is a little confusing.

Reply:
you're right, I wanted to change the return value based on if error code is present in inner response message or not, so I had chosen "Any". However, I changed my approach seeing that the default value for error code is already initialized to InvalidErrorCode. So using linq doesn't really lend any value here.